### ISSUE TITLE
Visualizing and Debugging Theano Shapes and Values

### ISSUE THREAD

**Issue by mustgoplay** [Theano/Theano#5465]
This another post for newbies like myself.  I have had a really difficult time trying to figure out how to tell what is going on inside my lasagna and keras   This is despite trying to digest the various debug pages in the documentation (which I recommend reading):

[Debugging Theano: FAQ and Troubleshooting](http://deeplearning.net/software/theano/tutorial/debug_faq.html)

[Using DebugMode](http://deeplearning.net/software/theano/tutorial/modes.html#using-debugmode)

[DebugPrint](http://deeplearning.net/software/theano/tutorial/printing_drawing.html#debug-print)

And despite sometimes being able to visualize the output of a keras layer:

[Function Suggestion - Visualize output of an input for a given layer/filter](https://github.com/fchollet/keras/issues/4884)

More often than not the debug information was simply graph information, or .eval() and get_value() failed to work and just gave errors and a text description of the tensor graph.

Thanks to an example from @patyork in the kerasteam slack, I've finally found a way to debug the code one step at a time.  The key was that instead of simply defining the function you want to debug in a few lines like in the documentation example, for me it was a lot easier to understand what is happening by using regular function format.

Disclaimer: In the examples below I use some keras functionality and I have pandas in there for when I use DataFrame to print large arrays but left regular prints for the example since a DataFrame print can't be 1D.

In the example below, the theano function to be debugged, theano_example, is just a regular python function however all the statements must be theano compatible:

```
import keras.backend as K
import theano.tensor as T
import pandas
import numpy as np
import theano

pandas.set_option('display.max_columns', None)
theano.config.exception_verbosity = 'high'

def theano_example(input1, input2):

    output1 = T.set_subtensor(input1[0], [5])
    output2 = input2[0]
    
    return output1, output2
        

input_1 = K.placeholder(shape=(None, 5)) # one example, of an 120x80 RGB image
input_2 = K.placeholder(shape=(None, 5)) # one example, of an 120x80 RGB image

theano_fn = theano_example(input_1, input_2)
fn = K.function(inputs=[input_1, input_2], outputs=theano_fn)

input_1 = ([0], [0.3], [0.5], [0.8], [0.2])
input_2 = np.ones((2,5))

out1, out2 = fn([input_1, input_2])
print('Input 1:')
print('shape: ', np.shape(input_1))
print(input_1)
print()
print('Input 2:')
print('shape: ', np.shape(input_2))
print(input_2)
print()
print('Output 1:')
print('shape: ', T.shape(out1).eval())
print(out1)
print()
print('Output 2:')
print('shape: ', T.shape(out2).eval())
print(out2)

```
The other advantage is that you can even reference functions within the function also using regular python format with theano compatible code.  Note in the code below theano_example calls theano_inner:

```
import keras.backend as K
import theano.tensor as T
import pandas
import numpy as np
import theano

pandas.set_option('display.max_columns', None)
theano.config.exception_verbosity = 'high'

def theano_inner(input):
    output = T.sum(input)
    
    return output

def theano_example(input1, input2):

    output1 = T.set_subtensor(input1[0], [5])
    output2 = theano_inner(input2)
    
    return output1, output2
        

input_1 = K.placeholder(shape=(None, 5)) # one example, of an 120x80 RGB image
input_2 = K.placeholder(shape=(None, 5)) # one example, of an 120x80 RGB image

theano_fn = theano_example(input_1, input_2)
fn = K.function(inputs=[input_1, input_2], outputs=theano_fn)

input_1 = ([0], [0.3], [0.5], [0.8], [0.2])
input_2 = np.ones((1,5))

out1, out2 = fn([input_1, input_2])
print('Input 1:')
print('shape: ', np.shape(input_1))
print(input_1)
print()
print('Input 2:')
print('shape: ', np.shape(input_2))
print(input_2)
print()
print('Output 1:')
print('shape: ', T.shape(out1).eval())
print(out1)
print()
print('Output 2:')
print('shape: ', T.shape(out2).eval())
print(out2)

```
In both examples I chose to have multiple outputs on purpose.  The way I have been using this to debug keras code is:

1) Start with the base example
2) Add 1 line (or a few until a significant change is made) at a time
3) Change the return ouputs for the inner and outer functions to be the tensor(s) I want to evaluate
4) Run the function and check the output(s) and their shapes
5) Rinse and repeat starting at step 2) until function is complete
6) Apply working procedure to the network

I hope this helps others.

Sincerely,
MGP

**Comment by Sentient07** [comment]
@nouiz can we have something like FAQs in DLT, and have these examples there? It will be very useful for someone who is new to theano to start with such walkthroughs. @mustgoplay great stuff!
